compiler/noirc_evaluator/src/ssa/opt/array_get_from_if_else_result.rs

```diff
@@ -0,0 +1,248 @@
+use crate::ssa::{
+    ir::{function::Function, instruction::Instruction, types::Type, value::Value},
+    ssa_gen::Ssa,
+};
+
+impl Ssa {
+    // Given an original IfElse instruction is this:
+    //
+    //     v10 = if v0 then v2 else if v1 then v3
+    //
+    // and a later ArrayGet instruction is this:
+    //
+    //     v11 = array_get v4, index v4
+    //
+    // we optimize the latter to this:
+    //
+    //     v12 = array_get v2, index v4
+    //     v13 = array_get v3, index v4
+    //     v14 = if v0 then v12 else if v1 then v13
+    #[tracing::instrument(level = "trace", skip(self))]
+    pub(crate) fn array_get_from_if_else_result_optimization(mut self) -> Self {
+        for function in self.functions.values_mut() {
+            optimize_array_get_from_if_else_result(function);
+        }
+
+        self
+    }
+}
+
+fn optimize_array_get_from_if_else_result(function: &mut Function) {
+    let block = function.entry_block();
+    let dfg = &mut function.dfg;
+    let instructions = dfg[block].take_instructions();
+
+    for instruction_id in instructions {
+        // Only apply this optimization to ArrayGet
+        let Instruction::ArrayGet { array, index } = &dfg[instruction_id].clone() else {
+            dfg[block].instructions_mut().push(instruction_id);
+            continue;
+        };
+
+        // Don't optimize if the index is a constant (this is optimized later on in a different way)
+        if let Value::NumericConstant { .. } = &dfg[dfg.resolve(*index)] {
+            dfg[block].instructions_mut().push(instruction_id);
+            continue;
+        }
+
+        // Only if getting an array from a previous instruction
+        let Value::Instruction { instruction, .. } = &dfg[dfg.resolve(*array)] else {
+            dfg[block].instructions_mut().push(instruction_id);
+            continue;
+        };
+
+        // Only if that previous instruction is an IfElse
+        let Instruction::IfElse { then_condition, then_value, else_condition, else_value } =
+            &dfg[*instruction]
+        else {
+            dfg[block].instructions_mut().push(instruction_id);
+            continue;
+        };
+
+        let then_condition = *then_condition;
+        let then_value = *then_value;
+        let else_condition = *else_condition;
+        let else_value = *else_value;
+
+        let then_value_type = dfg.type_of_value(then_value);
+
+        // Only if the IfElse instruction has an array type
+        let Type::Array(element_types, _) = then_value_type else {
+            dfg[block].instructions_mut().push(instruction_id);
+            continue;
+        };
+
+        let element_types: &Vec<Type> = &element_types;
+
+        // Only if the array isn't of a tuple type (or a composite type)
+        if element_types.len() != 1 {
```

Comment:
I'm not sure how to handle this case. Without this, `insert_instruction_and_results` below returns something that has multiple results and I don't know how to move that on to the next instruction. But maybe this optimization wasn't intended for composite types?